cli/tools/upgrade.rs

```diff
@@ -41,6 +41,16 @@ pub async fn upgrade_command(
 
   let install_version = match version {
     Some(passed_version) => {
+      if canary
+        && !regex::Regex::new("^[0-9a-f]{40}$")?.is_match(&passed_version)
+      {
+        eprintln!("Invalid commit hash passed");
+        std::process::exit(1)
+      } else if !canary && semver_parse(&passed_version).is_err() {
+        eprintln!("Invalid semver passed");
+        std::process::exit(1)
```

Comment:
I get that this is copying what's already done in this file, but wouldn't it be better to return errors rather than exiting the process here? It would be easier to add some tests for these failing scenarios in that case too.

Reply:
not sure what you mean

Reply:
So instead of exiting the process, doing something along the lines of:

```ts
return Err(type_error("Invalid semver passed"));
```

Reply:
`type_error` is used more for ops because it specifies the TypeError class. don't see why we should use that for subcommands; it also would be a first for that usecase

Reply:
Then use `custom_error` or `anyhow!()`

Reply:
done.